File: backend/api/overrides.py
"""
VoiceGuard — backend/api/overrides.py
CRUD REST API for the SQLite audio-hash override cache.

Endpoints:
    GET    /api/overrides          — list all override entries
    POST   /api/overrides          — create / update an override
    DELETE /api/overrides/{hash}   — remove an override by SHA-256 hash
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

# Ensure backend root is on path
_BACKEND_DIR = Path(__file__).parent.parent.resolve()
if str(_BACKEND_DIR) not in sys.path:
    sys.path.insert(0, str(_BACKEND_DIR))

from db.database import upsert_override, list_overrides, delete_override

router = APIRouter(prefix="/api/overrides", tags=["Demo Overrides"])


# ---------------------------------------------------------------------------
# Request / response models
# ---------------------------------------------------------------------------

class OverrideCreateRequest(BaseModel):
    sha256_hash: str = Field(..., description="SHA-256 hex digest of the audio file bytes")
    filename: str = Field("", description="Original filename for display")
    spoof_probability: float = Field(0.97, ge=0.0, le=1.0)
    prediction: str = Field("spoof", pattern="^(spoof|real)$")
    risk_level: str = Field("high", pattern="^(low|medium|high)$")
    risk_score: float = Field(0.97, ge=0.0, le=1.0)
    recommendation: str = Field(
        "🚨 AI-generated voice detected (Demo Override). Do NOT proceed — verify through a trusted channel.",
    )
    note: str = Field("", description="Optional free-form label, e.g. 'ElevenLabs clone'")


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@router.get("/", summary="List all demo overrides")
async def get_overrides() -> JSONResponse:
    """Return all stored audio-hash overrides (newest first)."""
    return JSONResponse(content=list_overrides())


@router.post("/", summary="Add or update a demo override")
async def create_override(req: OverrideCreateRequest) -> JSONResponse:
    """
    Create or replace the override for a given SHA-256 hash.
    When this hash is uploaded to /api/analyze, the stored scores are
    returned immediately without running AI inference.
    """
    upsert_override(
        sha256_hash=req.sha256_hash,
        filename=req.filename,
        spoof_probability=req.spoof_probability,
        prediction=req.prediction,
        risk_level=req.risk_level,
        risk_score=req.risk_score,
        recommendation=req.recommendation,
        note=req.note,
    )
    return JSONResponse(
        content={
            "status": "ok",
            "sha256_hash": req.sha256_hash,
            "message": f"Override saved for '{req.filename or req.sha256_hash[:12]}…'",
        }
    )


@router.delete("/{sha256_hash}", summary="Remove a demo override")
async def remove_override(sha256_hash: str) -> JSONResponse:
    """Delete the override for the given SHA-256 hash."""
    deleted = delete_override(sha256_hash)
    if not deleted:
        raise HTTPException(status_code=404, detail="Override not found.")
    return JSONResponse(content={"status": "ok", "sha256_hash": sha256_hash})

File: frontend/src/api/overrides.ts
/**
 * VoiceVeritas — src/api/overrides.ts
 * API client for the SQLite demo-override cache.
 * Endpoints: GET/POST/DELETE /api/overrides
 */

const BACKEND_URL = import.meta.env.VITE_BACKEND_URL ?? "http://localhost:8000";

export interface OverrideEntry {
  sha256_hash: string;
  filename: string;
  spoof_probability: number;
  prediction: "real" | "spoof";
  risk_level: "low" | "medium" | "high";
  risk_score: number;
  recommendation: string;
  note: string;
  created_at: string;
}

export interface CreateOverrideRequest {
  sha256_hash: string;
  filename?: string;
  spoof_probability?: number;
  prediction?: "real" | "spoof";
  risk_level?: "low" | "medium" | "high";
  risk_score?: number;
  recommendation?: string;
  note?: string;
}

/** Fetch all stored overrides (newest first). */
export async function fetchOverrides(): Promise<OverrideEntry[]> {
  const res = await fetch(`${BACKEND_URL}/api/overrides`);
  if (!res.ok) throw new Error("Failed to fetch overrides.");
  return res.json() as Promise<OverrideEntry[]>;
}

/** Save an override for the given SHA-256 hash. */
export async function saveOverride(req: CreateOverrideRequest): Promise<void> {
  const body: CreateOverrideRequest = {
    spoof_probability: 0.97,
    prediction: "spoof",
    risk_level: "high",
    risk_score: 0.97,
    recommendation:
      "🚨 AI-generated voice detected (Demo Override). Do NOT proceed — verify through a trusted channel.",
    note: "",
    ...req,
  };
  const res = await fetch(`${BACKEND_URL}/api/overrides`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Unknown error" }));
    throw new Error((err as { detail: string }).detail ?? "Failed to save override.");
  }
}

/** Remove an override by its SHA-256 hash. */
export async function deleteOverride(sha256Hash: string): Promise<void> {
  const res = await fetch(
    `${BACKEND_URL}/api/overrides/${encodeURIComponent(sha256Hash)}`,
    { method: "DELETE" }
  );
  if (!res.ok) throw new Error("Failed to delete override.");
}
